### PR TITLE
Add code docs to the Analytics classes, make `get_order_statuses` non-internal

### PR DESCRIPTION
**Pull request by tomalec**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Add code docs to the Analytics classes and make `get_order_statuses` non-internal.

Document analytics classes and some methods. Add simple examples of how to use it. Give a brief "why?" and "when?" answers. Add links between related classes.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Review the code docs.
1. Generate [code reference](https://github.com/woocommerce/code-reference) for the latest release: `./deploy.sh -v --build-only -s 9.1.4`
2. Checkout this branch and build woocommerce.zip: 
```sh
git checkout dev/fix-phpdoc-code-reference
cd plugins/woocommerce
pnpm run build:zip
```
4. unzip file to the code-reference directory
5. Generate again from local changes ``./deploy.sh -v --build-only -s 9.1.4 --no-download``
6. Review generated pages:
 - `file:///path/to/code-reference/build/api/classes/Automattic-WooCommerce-Admin-API-Reports-GenericController.html`
 - `file:///path/to/code-reference/build/api/classes/Automattic-WooCommerce-Admin-API-Reports-OrderAwareControllerTrait.html`
 - `file:///path/to/code-reference/build/api/classes/Automattic-WooCommerce-Admin-API-Reports-DataStore.html`
 - `file:///path/to/code-reference/build/api/classes/Automattic-WooCommerce-Admin-API-Reports-GenericStatsController.html`
 - `file:///path/to/code-reference/build/api/classes/Automattic-WooCommerce-Admin-API-Reports-StatsDataStoreTrait.html`

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

**Comment by tomalec** [comment]
I was not 100% sure if we should make `OrderAwareControllerTrait` internal or public. 

Previously, those methods were internal (the entire Report\Controller class was internal). And I'd like to avoid exposing too much as then backward compatibility would limit our pace of improvements. 
But I think it's worth exposing at least `get_order_statuses` to 3PDs to make reports consistent in terms of which orders are taken into account. However, that method specifically marked as internal in https://github.com/woocommerce/woocommerce-admin/pull/8429 

@louwie17 do you remember why?

File: plugins/woocommerce/src/Admin/API/Reports/DataStore.php [hunk 1]
```diff
@@ -13,7 +13,46 @@
 use Automattic\WooCommerce\Admin\API\Reports\TimeInterval;
 
 /**
- * Admin\API\Reports\DataStore: Common parent for custom report data stores.
+ * Common parent for custom report data stores.
+ *
+ * We use Report DataStores to separate DB data retrieval logic from the REST API controllers.
+ *
+ * Handles caching, data normalization, intervals-related methods, and other common functionality.
+ * So, in your custom report DataStore class that extend this class
```
Comment (ianlin) [inline plugins/woocommerce/src/Admin/API/Reports/DataStore.php]: Minor typo checked by chatgpt: should be `extends`.

File: plugins/woocommerce/src/Admin/API/Reports/GenericController.php [hunk 2]
```diff
@@ -7,10 +7,47 @@
...
+ *     }
+ * }
+ * </code></pre>
+ *
+ * The above Controller will use {@see GenericQuery GenericQuery}
+ * to get the data from a {@see DataStore data store} registered as `reports/my-thing`.
+ * (To change the Query class or data store name, override `construct_query` method).
+ *
+ * To use it, please register your controller in the `woocommerce_admin_rest_controllers`
+ * filter and initialize the `Rest_API` class in your plugin.
```
Comment (ianlin) [inline plugins/woocommerce/src/Admin/API/Reports/GenericController.php]: I was thinking that if I saw this for the first time, I might assume `Rest_API` is something specific to WooCommerce that I need to initialise.

**Review by ianlin — COMMENTED** [review]
Thanks for adding code docs. I haven't tested generating the code reference locally as I ran into a problem when building woocommerce.zip:

```
> wp --allow-root i18n make-pot . --include="woocommerce.php,client,i18n,includes,lib,packages,patterns,src,templates" --slug=woocommerce '--skip-audit'
Plugin file detected.
PHP Fatal error:  Allowed memory size of 536870912 bytes exhausted (tried to allocate 20480 bytes) in /Users/ianlin/dev/automattic/woocommerce/plugins/woocommerce/bin/composer/wp/vendor/mck89/peast/lib/Peast/Syntax/CommentsRegistry.php on line 172
```

Probably my local env issue, but weird thing is it failed when tried to allocate memory that is a lot lower than the allowed memory. Anyway I will keep trying it tomorrow.

**Comment by mikkamp** [comment]
> Probably my local env issue, but weird thing is it failed when tried to allocate memory that is a lot lower than the allowed memory. Anyway I will keep trying it tomorrow.

"Allowed memory size of 536870912 bytes exhausted (tried to allocate 20480 bytes)", means it tried to allocate 20k on top of the already exhausted memory.

In [other places we specifically tell it to set a higher memory limit](https://github.com/woocommerce/google-listings-and-ads/blob/2.8.2/package.json#L102) than the one you have.

**Comment by tomalec** [comment]
Thanks, @ianlin, for the review. I addressed all your comments.

> Thanks for adding code docs. I haven't tested generating the code reference locally as I ran into a problem when building woocommerce.zip:


You can also run the `code-reference` build for a released package (`./deploy.sh -v --build-only -s 9.1.4`), then replace the source files in the `woocommerce` sub-directory

File: plugins/woocommerce/src/Admin/API/Reports/FilteredGetDataTrait.php [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+ *     // Use the trait.
+ *     use FilteredGetDataTrait;
+ *     // Provide all the necessary properties and methods for a regular DataStore.
+ *     // ...
+ * }
+ * </code></pre>
+ *
+ * @see DataStore
+ */
+trait FilteredGetDataTrait {
```
Comment (ianlin) [inline plugins/woocommerce/src/Admin/API/Reports/FilteredGetDataTrait.php:29]: I don’t see this trait being used anywhere in the PR. Is it intended for an upcoming PR, or was it used in a different PR and is showing here due to merging?

File: plugins/woocommerce/src/Admin/API/Reports/DataStore.php [hunk 4]
```diff
@@ -48,9 +48,16 @@
  * </code></pre>
  *
  * Please use the `woocommerce_data_stores` filter to add your custom data store to the list of available ones.
- * Then, your store could be accessed by Query classes ({@see GenericQuery::get_data() GenericQuery::get_data()})
+ * Then, your store could be accessed by Controller classes ({@see GenericController::get_datastore_data() GenericController::get_datastore_data()})
  * or using {@link \WC_Data_Store::load() \WC_Data_Store::load()}.
```
Comment (ianlin) [inline plugins/woocommerce/src/Admin/API/Reports/DataStore.php:52]: Checking the generated report and I found the `@see` and `@link` here were not being transformed to the actual link. Would this related to the https://github.com/woocommerce/woocommerce/pull/50748#discussion_r1723766187 you added that we might need to fix the code reference?

Screenshot 2024-08-21 at 12 18 51

**Review by ianlin — COMMENTED** [review]
@tomalec Thanks for the extra changes, also thanks @mikkamp for the explanation and tip about exhausted memory.

I left some questions below that we can have further discussion.

File: plugins/woocommerce/src/Admin/API/Reports/FilteredGetDataTrait.php (hunk 3 above)
Comment (tomalec) [reply plugins/woocommerce/src/Admin/API/Reports/FilteredGetDataTrait.php:29]: The intent is to use in for the new classes and for 3PDs.

The existing controllers will still use `Query` classes for backward compatibility

File: plugins/woocommerce/src/Admin/API/Reports/DataStore.php (hunk 4 above)
Comment (tomalec) [reply plugins/woocommerce/src/Admin/API/Reports/DataStore.php:52]: yes, exactly. I didn't manage the code reference to work with them. I think we can fix that later.

File: plugins/woocommerce/src/Admin/API/Reports/FilteredGetDataTrait.php (hunk 3 above)
Comment (tomalec) [reply plugins/woocommerce/src/Admin/API/Reports/FilteredGetDataTrait.php:29]: See the discussions at https://github.com/woocommerce/woocommerce/pull/50749 as it was reviewed there.

Sorry, for merging it before review of this one is finished :(

**Review by ianlin — APPROVED** [review]
Thanks for answering my questions. It's good to go, LGTM.

**Pull request merged**